runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -145,26 +153,32 @@ public Token appendCheckpoint() {
      * This API should not be directly invoked.
      *
      *  @param snapshotTimestamp snapshot at which the checkpoint is taken.
+     *  @param streamTail tail of the stream to checkpoint at snapshot time.
      */
     @VisibleForTesting
-    public Token appendCheckpoint(Token snapshotTimestamp) {
+    public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
         long start = System.currentTimeMillis();
 
-        // If snapshotTimestamp (stream tail) is a negative value
+        // If stream tail is a negative value
         // we need to enforce a hole to ensure the start log address
         // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
         // Having a negative value will prevent trimming from happening.
-        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+        if (!Address.isAddress(streamTail)) {
             snapshotTimestamp = forceHole();
+            streamTail = snapshotTimestamp.getSequence();
```

Comment:
This is inconsistent. If you look at `CheckpointWriter::forceHole` it can return a different token than the one passed to `AddressSpaceView::write` because internally it uses `layoutHelper` which can reset the token.

Reply:
Since its a hole, why not just use `AddressSpace::read` directly and let if fill the hole?

Reply:
I think my first comment is incorrect. The addressspace write doesn't change the token, so `forceHole` should only fail with a OverwriteException/Trimmed Exception

Reply:
Good point! Done.

Reply:
But what if there is a sequencer failover (between acquiring the token for the hole and writing)... and this same position ends up being a hole but for a different stream, then the validateStateOfWrittenEntry succeeds cause it's a hole, that would be an issue, right? At the end we always have a problem because token request and write/read for hole filling is not performed atomically. We can always have regressions that lead to data loss.

Reply:
Never mind, the epoch validation that we do afterwards ensures this won't happen.